### DISCUSSION TITLE
[run/compile]IvyDeps shall be transitive to the nested test module like ivyDeps

### DISCUSSION THREAD

**Discussion by b-gyula** [com-lihaoyi/mill#1463] [Ideas]
Like in [gradle](https://docs.gradle.org/current/userguide/declaring_dependencies.html): If some dependency needed for compile the artifact (like a macro library), it's most probably needed for compile the tests as well. 
The same applies for run: when an artifact needs a dependency to be able to run, that is certainly needed to run a test also. E.g. If a DB driver needed for an app to run, then test won't be able to run it without it for sure.

**Comment by lefou** [comment]
I definitely disagree for `compileIvyDeps`. It's their exact purpose to be compile-time only and not transitive.

I personally like the non-transitive behavior of  `runIvyDeps` in Mill very much, as it gives more power in rather large projects. Working this way is much more pleasant (to me) than was all that fiddling with excludes in other build tools.

**Reply by Sailsman63** [reply]
So, for the average mill build it's expected to have a lot of this in the buildfile?

````
object foo extends JavaModule {
   object test extends Tests with TestModule.JUnit4 {
     def runClasspath = foo.runClasspath() // as the very most basic. Almost certain to crash without it.
  }
}
````

**Reply by lihaoyi** [reply]
I've never written that line in my 4 years of creating and using Mill builds, and a quick github search shows most others have not either. I don't expect most people to need such a line. 

Perhaps you could give us more information on your specific situation and what you are trying to achieve, what the problems are, then we can best figure out what the ideal course or action is. It's not obvious to me whether transitive or non-transitive is better, but currently it is non transitive, and we'd need a convincing case to swap that behavior

**Reply by Sailsman63** [reply]
I'm not entirely certain which would be a better default, either. Still learning here.

I've actually had to add this for two different modules so far. In both cases, the class (Java) under test depends on other classes in the project. The tests are a little broader than unit, so `foo` needs a `bar` and `baz` to operate on. `bar` in turn has dependencies on our 3rd-party dependencies, which will not resolve unless I add the above.

**Reply by Sailsman63** [reply]
Okay, a little more detail, since my attempts to work around this have all crashed and burned:

I've got a set of `JavaModules` that all have a common set of dependencies:

```
trait Common extends JavaModule {
   def ivyDeps = Agg(...)

  def unmanagedClasspath = T {
    Agg(...)
  }
}

object Module1 extends Common {
...
  object test extends Tests with TestModule.JUnit4 {
  }
}
object Module2 extends Common {}

```
The tests, of course, have to instantiate classes in Module1. (This is a plain Java application, and a lot of the code is legacy. Not as many pure function - type constructs as you may be used to.) Some of the code under test relies on our third-party dependency being available. This is pretty tightly integrated, and nearly ubiquitous. (When a large chunk of the codebase is an M(VC) editor with a somewhat thin Model component, it's natural for your UI library to have its nose in everything)

I can't mix in `Common`, because it and `TestModule.JUnit4` keep stepping on each other. If I do:

```
object test extends Tests with Common with TestModule.JUnit4
```

The JUnit4 module overrides the `runClasspath`, and the behavior is as in the main case. The testcases compile, but can't run. If I do:

```
object test extends Tests with TestModule.JUnit4 with Common
```

The Common module will take over the compileClasspath, and the tests won't compile because the compiler doesn't see the JUnit library.

(This probably seems obvious to you, and it does make sense when I think about it - Scala noob here, haven't internalized how this works yet.)

How would you suggest configuring for this set of tests?

**Reply by lefou** [reply]
You may want to avoid to reset the `ivyDeps` in `Common`.

```scala
def ivyDeps = T { super.ivyDeps() ++ Agg(...) }
```

**Reply by Sailsman63** [reply]
Good catch in general. It does not fix my main issue here, though.

- It does allow me to *compile* tests in either order.
- Tests still won't run without directly setting the `runClasspath`. It seems that the testrunner, somewhere, ends up ignoring the `unmanagedClasspath`

**Reply by lefou** [reply]
You can inspect it's dependencies with `mill inspect Module1.test.runClasspath` and the result with `mill show Module1.test.runClasspath`.

**Reply by Sailsman63** [reply]
Okay, this is weird. The dependency that keeps getting missed is, indeed, present on `Module1.test.runClasspath` (via `test.upstreamAssemblyClasspath` < `test.unmanagedClasspath` which is pulled in from `Common`)

Nevertheless, it still fails to run unless I explicitly include `Module1.upstreamAssemblyClasspath`. Comparing `mill show Module1.test.runClasspath` from both configurations shows the unmanaged files being added a second time, with different `ref:<hash>` values.

**Reply by lefou** [reply]
What is `mill plan` and `mill visualizePlan` looking like? There must be something (not obvious too me) in your build. You can also use `mill path` to show the dependency chain between two tasks. The different hashes are definitely weird. Do you create the `PathRef` before you actually touch/modify that path? Creating the PathRef should be the last operation after creating the file it refers to.

**Reply by Sailsman63** [reply]
At bit like a spider web? Both do show *all* of the dependent segments being part of the tree in the non-working config. When I add the lines that make it work, that section of the task tree is duplicated.

The file that keeps getting missed is checked into the source repository, and added to the listing with a basic `PathRef(millSourcePath / os.up / "lib" / "filename.jar")`. When I run `mill show <Target>.unmanagedClasspath` the correct absolute path is shown for this file.

If you think a broader look at the build would help, it's FOSS. The mill build is being worked on/discussed on [this PR:](https://github.com/ArtOfIllusion/ArtOfIllusion/pull/275). As of commit 01c9989 (today), the lines that I have to add the get the tests to work are 51 and 61 in `build.sc`. If you comment them out, you should get the `ClassNotDefined Error` for a class in `buoy.widget` (it exists in the file `lib/buoy.jar`)

**Reply by lefou** [reply]
Interesting. I guess I'd like to try to test your project, but can't promise anything, as my time is fairly limited.

**Reply by lefou** [reply]
Can you post the concrete mill commands which fail? Also, are there any other requirements besides checking out and running Mill?

**Reply by lefou** [reply]
Hm, this whole thread was initially about transitivity of `ivyDeps`. But from a quick glance at your linked PR (https://github.com/ArtOfIllusion/ArtOfIllusion/pull/275) you're using `unmanagedClasspath` for almost all dependencies. Please beware, that `unmanagedClasspath` is not transitive. I added a comment to your PR.

**Reply by lefou** [reply]
It's also very common when working with non-public jars, to provision them locally into the local Ivy or Maven repository and then just use them via `ivyDeps`. `mill.scalalib.publish.LocalIvyPublisher` or `mill.scalalib.publish.LocalM2Publisher` is your friend. It's in most cases ok to just push empty doc and source jars, if you don't have them.

**Reply by Sailsman63** [reply]
Yeah, sorry that I've sort of strayed from the original topic. I thought they were thematically the same when I started.

I will look into the local publisher options - I'm aware that unmanaged dependencies are not ideal, but I kind of need to get the build working before I can think too hard about upgrading dependencies. (Our existing legacy ant build is *entirely* unmanaged dependencies, and one of the motivators in moving to something else is ease of transitioning to managed dependencies. This build is already an improvement in that regard)
